File: basic_sort/basic_sort.go
package main

import "fmt"

func bubbleSort(arr []int) []int {
	n := len(arr)
	for i := n - 1; i > 0; i-- {
		swapped := false
		for j := 0; j < i; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
	return arr
}

func selectionSort(arr []int) []int {
	n := len(arr)
	for i := 0; i < n-1; i++ {
		minIndex := i
		for j := i + 1; j < n; j++ {
			if arr[j] < arr[minIndex] {
				minIndex = j
			}
		}
		if i != minIndex {
			arr[i], arr[minIndex] = arr[minIndex], arr[i]
		}
	}
	return arr
}

func insertionSort(arr []int) []int {
	for i := 1; i < len(arr); i++ {
		currentValue := arr[i]
		j := i - 1
		for j >= 0 && arr[j] > currentValue {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = currentValue
	}
	return arr
}

func main() {
	arr := []int{5, 8, 1, 6, 7, 4, 3}
	
	fmt.Println(bubbleSort(arr))
	fmt.Println(selectionSort(arr))
	fmt.Println(insertionSort(arr))
}

File: basic_sort/basic_sort.py
def bubble_sort(lst):
    for i in range(len(lst) - 1, 0, -1):
        swapped = False
        for j in range(i):
            if lst[j] > lst[j + 1]:
                lst[j], lst[j + 1] = lst[j + 1], lst[j]
                swapped = True
        if not swapped:
            break
    return lst

def selection_sort(lst):
    for i in range(len(lst) - 1):
        min_index = i
        for j in range(i + 1, len(lst)):
            if lst[j] < lst[min_index]:
                min_index = j
        if i != min_index:
            lst[i], lst[min_index] = lst[min_index], lst[i]
    return lst

def insertion_sort(lst):
    for i in range(1, len(lst)):
        current_value = lst[i]
        j = i - 1
        while j >= 0 and current_value < lst[j]:
            lst[j + 1] = lst[j]
            j -= 1
        lst[j + 1] = current_value
    return lst

print(bubble_sort([5, 8, 1, 6, 7, 4, 3]))
print(selection_sort([5, 8, 1, 6, 7, 4, 3]))
print(insertion_sort([5, 8, 1, 6, 7, 4, 3]))
